modal/sandbox.py

```diff
@@ -26,17 +27,33 @@
 
 
 class _LogsReader:
-    """Provides an interface to buffer and fetch logs from a sandbox stream (`stdout` or `stderr`)."""
+    """Provides an interface to buffer and fetch logs from a sandbox stream (`stdout` or `stderr`).
+
+    As an asynchronous iterable, the object supports the async for statement.
+    **Usage**
+
+    ```python
+    sandbox = stub.app.spawn_sandbox("bash", "-c", "while true; do echo foo; sleep 1; done)
+    async for message in sandbox.stdout:
+        print(f"Message: {message}")
+    ```
+    """
 
     def __init__(self, file_descriptor: int, sandbox_id: str, client: _Client) -> None:
         """mdmd:hidden"""
 
         self._file_descriptor = file_descriptor
         self._sandbox_id = sandbox_id
         self._client = client
+        self._stream = None
+        self._last_log_batch_entry_id = ""
+        # Whether the reader received an EOF. Once eof is True, it returns
+        # an empty string for any subsequent reads (including async for)
+        self.eof = False
```

Comment:
The semantic for `LogsReader` in this PR is based on `asyncio.subprocess`'s semantic.

If we perform `reader.read()` after an `async for` (or vice versa), the latter will not read anything. This is because the reader has already encountered an `eof`. 

This behavior is shown in the example below:

```

import asyncio

async def read_output(process):
    async for line in process.stdout:
        print(f'Output: {line.decode().strip()}')

async def main():
    process = await asyncio.create_subprocess_exec(
        'bash', '-c', 'echo foo', 
        stdout=asyncio.subprocess.PIPE
        )
    async for line in process.stdout:
        print(f'Output: {line.decode()}')
    await read_output(process)

    first = await process.stdout.read()
    print(f"First: {first}")

    second = await process.stdout.read()
    print(f"Second: {second}")

asyncio.run(main())
```

Reply:
Before this PR, calling `read` twice would return the contents of the entire stream. Now, the second read would return an empty string.